### PR TITLE
Logging Doc

### PR DESCRIPTION
**Pull request by AChoffat**
Add documentation for logging part

### CODE REVIEW

**Comment by AMontagu** [comment]
I thought we agree on splitting logging into 2 parts. One in the Features folders that explain the logging system and one in the how to section explaining the way to add extra context ?
Did @legau asked change on that ?

**Comment by AChoffat** [comment]
I think it's me who didn't understand that

File: docs/features/logging.rst [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+Description
+-----------
+
+Django-Socio-GRPC has a built-in way to handle the logging of the errors of your services.
+    
+Usage
+-----
+
+Django-Socio-GRPC uses a system of logging based on the one used by Django.
```
Comment (AMontagu) [inline docs/features/logging.rst:12]: put the link to the django doc directly

File: docs/features/logging.rst [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+==========
+Formatters
+==========
+
+Ultimately, a log record needs to be rendered as text. Formatters describe the exact format of that text. A formatter usually consists of a Python formatting string containing LogRecord attributes; however, you can also write custom formatters to implement specific formatting behavior.
+
+
+Here's an example of a Logging setup that uses all four of these objects.
+
+.. code-block:: python
```
Comment (AMontagu) [inline docs/features/logging.rst:59]: this example may not implement solution or should not be in upur doc as junior dev can have tendancies to copy/paste example directly in code and this example does not just work out of the box.
What your opinion on that @legau

File: docs/features/logging.rst [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+=====================
+Logging GRPC Requests
+=====================
+
+Django-Socio-GRPC by default log requests only when something goes wrong just like Django does.
+    - If your service raises a GRPCException, it will result by default to a Warning.
+      You can define your own GRPCException and set the log_level as you wish to change this.
+    - If your service raises another Exception, it will result in an Error.
+    - Lastly, if you want to log requests even if they are ok, you can launch django in DEBUG mode
```
Comment (AMontagu) [inline docs/features/logging.rst:121]: replace by:
- Logging incoming request is only activated when in DEBUG mode or if LOG_OK_RESPONSE settings is set to True

File: docs/features/logging.rst [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    LOGGING = {
+        "version": 1,
+        "disable_existing_loggers": False,
+        "filters": {"require_debug_false": {"()": "django.utils.log.RequireDebugFalse"}},
+        "formatters": {
+            "classic": {
+                "format": "[django]-[%(levelname)s]-[%(asctime)s]-[%(name)s:%(lineno)s] %(message)s"
+            },
+            "fmt": {
+                "format": 'level=%(levelname)s name=%(name)s line=%(pathname)s:%(lineno)s message="%(message)s" socio_service_name="%(socio_service_name)s" socio_action="%(socio_action)s" time=%(asctime)s levelno=%(levelno)s socio_usermanagement_uuid="%(socio_usermanagement_uuid)s" socio_user_roles="%(socio_user_roles)s" socio_client_id="%(socio_client_id)s" funcName=%(funcName)s',
```
Comment (AMontagu) [inline docs/features/logging.rst:141]: you have specific socio context in this example. I think fmt format should not be documented in the logging features doc

File: docs/features/logging.rst [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+            },
+        },
+        "handlers": {
+            "null": {"level": "DEBUG", "class": "logging.NullHandler"},
+            "console": {
+                "level": logging.DEBUG if DEBUG else logging.INFO,
+                "class": "logging.StreamHandler",
+                "stream": sys.stdout,
+                "formatter": "fmt",
```
Comment (AMontagu) [inline docs/features/logging.rst:150]: use classic formatter for people to understand better

File: docs/how-to/add-extra-context-to-logging.rst [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+        logging.setLogRecordFactory(record_factory)
+
+In this example it will enable you to define formatters like this for example.
+
+.. code-block:: python
+    "formatters": {
+        "django_socio_grpc_formatter": {
+            "format": "{levelname} {grpc_service_name} {grpc_action} {grpc_user_pk} {message}",
+            "style": "{",
```
Comment (AMontagu) [inline docs/how-to/add-extra-context-to-logging.rst:69]: why style is needed here ?

File: docs/how-to/add-extra-context-to-logging.rst [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            return record
+
+        logging.setLogRecordFactory(record_factory)
+
+In this example it will enable you to define formatters like this for example.
+
+.. code-block:: python
+    "formatters": {
+        "django_socio_grpc_formatter": {
+            "format": "{levelname} {grpc_service_name} {grpc_action} {grpc_user_pk} {message}",
```
Comment (AMontagu) [inline docs/how-to/add-extra-context-to-logging.rst:68]: may be better to just add the extra context to the classic format for the example

File: docs/how-to/add-extra-context-to-logging.rst [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    extra_context = {
+        "grpc_service_name": service.get_service_name(),
+        "grpc_action": service.action,
+    }
+    if hasattr(service.context, "user") and hasattr(service.context.user, "pk"):
+        extra_context["grpc_user_pk"] = service.context.user.pk
+    return extra_context
+
+
+    def set_log_record_factory():
```
Comment (AMontagu) [inline docs/how-to/add-extra-context-to-logging.rst:35]: you don't explain how you have to call this method.
For information this method should be called before any log using the format and this should be explicitly documented IMO

File: docs/settings.rst [hunk 9]
```diff
@@ -235,3 +235,12 @@ For a standard configuration, you might have:
 This means that when the framework encounters metadata, it knows to look for a ``HEADERS``
 key to retrieve headers, a ``PAGINATION`` key to fetch pagination data, and a ``FILTERS`` key
 for filtering details.
+
+LOG_OK_RESPONSE
+^^^^^^^^^^^^^^^
+
+This setting enables the logging of requests that return an OK.
```
Comment (AMontagu) [inline docs/settings.rst:242]: should add here that the defaut behavior depend on the DEBUG settings if setting is False
And that defautl is False

File: docs/how-to/add-extra-context-to-logging.rst [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+-----
+
+=====================
+Logging GRPC Services
+=====================
+
+You'll want to be able to log information about your grpc services.
+Django-Socio-GRPC is built-in with existing loggers to get information from GRPC services
+
+If you want to add information to your logs from grpc. You will need to define these two functions and call set_log_record_factory in your settings.
```
Comment (AMontagu) [inline docs/how-to/add-extra-context-to-logging.rst:19]:
```suggestion
If you want to add information to your logs from grpc. You will need to define these two functions and call set_log_record_factory in your settings before any log using your formatter is called.
```

**Comment by AMontagu** [comment]
waiting for last comment and @legau review and okay for me

File: docs/how-to/add-extra-context-to-logging.rst [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    """
+    It allow logs to have extra data about the current context of the log. Used especially for tracing system.
+    """
+    extra_context = {
+        "grpc_service_name": service.get_service_name(),
+        "grpc_action": service.action,
+    }
+    if hasattr(service.context, "user") and hasattr(service.context.user, "pk"):
+        extra_context["grpc_user_pk"] = service.context.user.pk
+    return extra_context
```
Comment (legau) [inline docs/how-to/add-extra-context-to-logging.rst:32]: indentation ?

File: docs/how-to/add-extra-context-to-logging.rst [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        old_factory = logging.getLogRecordFactory()
+
+        def record_factory(*args, **kwargs):
+            from django_socio_grpc.services.servicer_proxy import get_servicer_context
+
+            servicer_ctx = get_servicer_context()
+
+            record = old_factory(*args, **kwargs)
+
+            record.grpc_service_name = ""
```
Comment (legau) [inline docs/how-to/add-extra-context-to-logging.rst:49]: comment that we set a default so the formatter will not crash when this data is missing

**Review by legau — CHANGES_REQUESTED** [review]

File: docs/features/logging.rst [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+
+Description
+-----------
+
+Django-Socio-GRPC has a built-in way to handle the logging of the errors of your services.
+    
+Usage
+-----
+
+Django-Socio-GRPC uses a system of logging based on the one used by Django. `Documentation <https://docs.djangoproject.com/fr/4.2/topics/logging/#topic-logging-parts-loggers>`_.
```
Comment (legau) [inline docs/features/logging.rst:14]:
```suggestion
Django-Socio-GRPC uses a system of logging based on the one used by Django. `Documentation <https://docs.djangoproject.com/en/4.2/topics/logging/#topic-logging-parts-loggers>`_.
```

File: docs/features/logging.rst [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+-----
+
+Django-Socio-GRPC uses a system of logging based on the one used by Django. `Documentation <https://docs.djangoproject.com/en/4.2/topics/logging/#topic-logging-parts-loggers>`_.
+
+
+=====================
+Logging GRPC Requests
+=====================
+
+Django-Socio-GRPC by default log requests only when something goes wrong just like Django does.
```
Comment (legau) [inline docs/features/logging.rst:21]: Specify which logger

**Pull request merged**